File: src/test/java/com/github/anddd7/algorithm/sort/TopologicalSortingTest.java
package com.github.anddd7.algorithm.sort;

import static java.util.Collections.emptyList;
import static java.util.Collections.singletonList;
import static java.util.Comparator.comparingInt;
import static org.assertj.core.api.AssertionsForInterfaceTypes.assertThat;
import static org.junit.jupiter.api.Assertions.assertThrows;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;
import org.junit.jupiter.api.TestInstance.Lifecycle;

@TestInstance(Lifecycle.PER_CLASS)
class TopologicalSortingTest {

  private List<Task> topologicalDAG = new ArrayList<>();
  private List<Task> circleDependencies = new ArrayList<>();

  @BeforeAll
  void setUp() {
    topologicalDAG.add(new Task(1, emptyList()));
    topologicalDAG.add(new Task(2, singletonList(1)));
    topologicalDAG.add(new Task(3, singletonList(1)));
    topologicalDAG.add(new Task(4, singletonList(2)));
    topologicalDAG.add(new Task(5, Arrays.asList(3, 4)));
    topologicalDAG.add(new Task(6, singletonList(1)));
    topologicalDAG.add(new Task(7, Arrays.asList(1, 5)));
    topologicalDAG.add(new Task(8, Arrays.asList(1, 2, 3)));
    topologicalDAG.add(new Task(9, singletonList(8)));
    topologicalDAG.add(new Task(10, emptyList()));
    topologicalDAG.add(new Task(11, emptyList()));
    topologicalDAG.add(new Task(12, singletonList(10)));

    circleDependencies.add(new Task(1, emptyList()));
    circleDependencies.add(new Task(2, singletonList(3)));
    circleDependencies.add(new Task(3, singletonList(4)));
    circleDependencies.add(new Task(4, singletonList(2)));
  }

  @Test
  void sorting_successful() {
    List<Task> result = TopologicalSorting.sort(
        topologicalDAG,
        e -> e.id, e -> e.dependencies,
        comparingInt(o -> o.id)
    );
    List<Integer> sequence = result.stream().map(task -> task.id).collect(Collectors.toList());

    assertThat(sequence)
        .containsSequence(
            1, 10, 11, 2, 3, 6, 12, 4, 8, 5, 9, 7
        );
  }

  @Test
  void sorting_failed() {
    assertThrows(
        IllegalArgumentException.class,
        () -> TopologicalSorting.sort(circleDependencies, e -> e.id, e -> e.dependencies, comparingInt(o -> o.id))
    );
  }

  static class Task {

    Integer id;
    List<Integer> dependencies;

    Task(Integer id, List<Integer> dependencies) {
      this.id = id;
      this.dependencies = dependencies;
    }
  }
}

File: kotlin/leetcode-kotlin/src/test/kotlin/com/github/anddd7/leetcode/sort/TopologicalSortingTest.kt
package com.github.anddd7.leetcode.sort

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows

internal class TopologicalSortingTest {

    private val topologicalDAG: List<Pair<Int, List<Int>>> = mutableListOf(
        Pair(1, emptyList()),
        Pair(2, listOf(1)),
        Pair(3, listOf(1)),
        Pair(4, listOf(2)),
        Pair(5, listOf(3, 4)),
        Pair(6, listOf(1)),
        Pair(7, listOf(1, 5)),
        Pair(8, listOf(1, 2, 3)),
        Pair(9, listOf(8)),
        Pair(10, emptyList()),
        Pair(11, emptyList()),
        Pair(12, listOf(10))
    )

    @Test
    fun `topological sorting with generic type`() {
        val result = TopologicalSorting(
            topologicalDAG.shuffled(),
            { first }, { second },
            Comparator.comparingInt { it.first }
        ).sort()
        val resultIds = result.map { it.first }

        assertThat(resultIds)
            .containsSequence(1, 10, 11, 2, 3, 6, 12, 4, 8, 5, 9, 7)
    }

    private val circleDependencies: List<Pair<Int, List<Int>>> = mutableListOf(
        Pair(1, emptyList()),
        Pair(2, listOf(3)),
        Pair(3, listOf(4)),
        Pair(4, listOf(2))
    )

    @Test
    fun `throw IllegalArgumentException while have circle dependencies`() {
        val exception = assertThrows<IllegalArgumentException> {
            TopologicalSorting(
                circleDependencies.shuffled(),
                { first }, { second },
                Comparator { o1, o2 -> o1.first - o2.first }
            )
                .sort()
        }
        assertThat(exception)
            .hasMessage("Cannot sort these elements, because of circle dependencies between: [2, 3, 4]")
    }
}
